src/os/bluestore/BlueStore.cc

```diff
@@ -532,6 +532,184 @@ ostream& operator<<(ostream& out, const BlueStore::Buffer& b)
+  uint64_t lookup_end_offset = ROUND_UP_TO(end_offset, min_alloc_size);
+
+  dout(30) << __func__ << " (hex): [" << std::hex
+           << lookup_start_offset << ", " << lookup_end_offset 
+           << ")" << std::dec << dendl;
+
+  for (auto it = extent_map.seek_lextent(lookup_start_offset);
+       it != extent_map.extent_map.end() &&
+         it->logical_offset < lookup_end_offset;
+       ++it) {
+    uint64_t alloc_unit_start = it->logical_offset / min_alloc_size;
+    uint64_t alloc_unit_end = (it->logical_end() - 1) / min_alloc_size;
+
+    dout(30) << __func__ << " " << *it
+             << "alloc_units: " << alloc_unit_start << ".." << alloc_unit_end
+             << dendl;
+
+    Blob* b = it->blob.get();
+
+    if (it->logical_offset >=start_touch_offset &&
+        it->logical_end() <= end_touch_offset) {
+      // Process extents within the range affected by 
+      // the current write request.
+      // Need to take into account if existing extents
+      // can be merged with them (uncompressed case)
+      if (!b->get_blob().is_compressed()) {
+        if (blob_info_counted && used_alloc_unit == alloc_unit_start) {
+	  --blob_info_counted->expected_allocations; // don't need to allocate
+                                                     // new AU for compressed
+                                                     // data since another
+                                                     // collocated uncompressed
+                                                     // blob already exists
+          dout(30) << __func__  << " --expected:"
+                   << alloc_unit_start << dendl;
+        }
+        used_alloc_unit = alloc_unit_end;
+        blob_info_counted =  nullptr;
+      }
+    } else if (b->get_blob().is_compressed()) {
+
+      // additionally we take compressed blobs that were not impacted
+      // by the write into account too
+      auto b_it =
+        affected_blobs.emplace(b, BlobInfo(b->get_referenced_bytes())).first;
+      BlobInfo& bi = b_it->second;
+      for (auto i = alloc_unit_start; i <= alloc_unit_end; ++i) {
+	if (!used_alloc_unit || i != used_alloc_unit  ) {
+          ++bi.expected_allocations;
+          dout(30) << __func__  << " ++expected_allocations:" << i << dendl;
+	  blob_info_counted =  &bi;
+        }
+	used_alloc_unit = i;
+      }
+
+      assert(it->length <= bi.referenced_bytes);
+      bi.referenced_bytes -= it->length;
+      dout(30) << __func__ << " affected_blob:" << *b
+               << " unref 0x" << std::hex << it->length
+               << " referenced = 0x" << bi.referenced_bytes
+               << std::dec << dendl;
+      // NOTE: we can't move specific blob to resulting GC list here
+      // when fake_ref_map is empty since subsequent extents might
+      // decrement its expected_allocation. 
+      // Hence need to enumerate all the extents first.
+      bi.collect_candidates.emplace_back(it->logical_offset, it->length);
```

Comment:
I wonder if it's better to skip building collection_candidates, and instead, if we decide this blob is toast, just iterate over the extent_map for the blob range.  It'll be hot in the CPU cache, and we'll avoid the allocations for the map.  Most of the time there won't be extents in that range to other blobs anyway, so we won't even be enumerating more elements than are in collect_candidates, I think?